File: Project/Simulator/Organisms/Plants/Plant.java
package Simulator.Organisms.Plants;

import Simulator.Organisms.Organism;
import Simulator.World;
import Simulator.GlobalSettings;
import Utils.RNG;
import Simulator.Tile;

import org.json.JSONObject;

public abstract class Plant extends Organism {

    private int reproduction_cooldown = 5;

    public Plant(int power, World world, Type type) {
        super(type, power, 0, world);
    }

    public Plant(JSONObject json, World world) {
        super(json, world);
    }

    @Override
    public void action() {
        if (!GlobalSettings.AI_REPRODUCE)
            return;
        if (RNG.roll(0, 100) < 5) {
            Tile newtile = tile.getRandomFreeNeighbour();
            if (newtile == null)
                return;
            Plant newPlant = construct();
            newPlant.skipTurn();
            world.addOrganism(newPlant, newtile);
            // this.setSpreadCooldown(5);
            // newPlant.setSpreadCooldown(5);

            world.addLog(getSymbol() + " is spreading!!");
        }
    }

    @Override
    public void collision(Organism other) {
    }

    @Override
    public boolean collisionReaction(Organism other) {
        return false;
    }

    public void setSpreadCooldown(int cooldown) {
        this.reproduction_cooldown = cooldown;
    }

    public int getSpreadCooldown() {
        return this.reproduction_cooldown;
    }

    public abstract Plant construct();

}

File: Project/Simulator/Organisms/Plants/Plant.py
import random

from abc import ABC, abstractmethod

from Simulator.Organisms.Organism import Organism
from Simulator.GlobalSettings import GlobalSettings


class Plant(Organism, ABC):
    def __init__(self, power, world, type, json=None):
        super().__init__(type, power, 0, world, json)
        self._reproduction_cooldown = 5

    def action(self):
        if not GlobalSettings.AI_REPRODUCE:
            return
        if random.randint(0, 100) < 5:
            if self._tile is not None:
                newtile = self._tile.getRandomFreeNeighbour()
                if newtile is None:
                    return
                newPlant = self.construct()
                if newPlant is None:
                    return
                newPlant.skipTurn()
                self._world.addOrganism(newPlant, newtile)

                self._world.addLog(self.getSymbol() + " is spreading!!")

    def collision(self, other):
        pass

    def collisionReaction(self, other):
        return False

    def setSpreadCooldown(self, cooldown):
        self._reproduction_cooldown = cooldown

    def getSpreadCooldown(self):
        return self._reproduction_cooldown

    @abstractmethod
    def construct(self):
        pass
